Documentation/teaching/so2/assign2-driver-uart.rst

```diff
@@ -0,0 +1,163 @@
+==========================
+Assignment 2 - Driver UART
+==========================
+
+-  Deadline: :command:`Wednesday, 21 April 2021, 23:00`
+
+Assignment's Objectives
+=======================
+
+*  consolidating the knowledge of device driver
+*  read hardware documentation and track the desired functionality in the documentation
+*  work with interrupts; use of non-blocking functions in interrupt context
+*  use of buffers; synchronization
+*  kernel modules with parameters
+
+Statement
+=========
+
+Write a kernel module that implements a driver for the serial port (`UART16550`).
+The device driver must support the two standard serial ports in a PC, `COM1` and `COM2` (`0x3f8` and `0x2f8`,
+in fact the entire range of `8` addresses `0x3f8-0x3ff` and `0x2f8-0x2ff` specific to the two ports).
+In addition to the standard routines (`open`, `read`, `write`, `close`),
+the driver must also have support for changing communication parameters using an `ioctl` operation (`UART16550_IOCTL_SET_LINE`).
+
+The driver must use interrupts for both reception and transmission to reduce latency and CPU usage time.
+`Read` and `write` calls must also be blocking. :command:`Assignments that do not meet these requirements will not be considered.`
+It is recommended that you use a buffer for read and another buffer for write buffer for each serial port in the driver.
+
+A blocking read call means that the read routine called from the user-space will be blocked until :command:`at least` one byte is read
+(the read buffer in the kernel is empty and no data can be read).
+A blocking write call means that the write routine called from the user-space will be blocked until :command:`at least` one byte is written
+(the write buffer in the kernel is full and no data can be written).
+
+Buffers Scheme
+--------------
+
+.. image:: ../img/buffers-scheme.png
+
+Data transfer between the various buffers is a `Producer-Consumer <https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem>`__ problem. Example:
+
+-   if it is written from a process to the device, the process is the producer and the device is the consumer; the process will block until there is at least one free space in the consumer's buffer
```

Comment:
rephrase: - if a process writes to the device the process is the producer...
same for the next sentence, with the read case